lib/java/com/google/android/material/bottomsheet/BottomSheetDialog.java

```diff
@@ -112,6 +115,49 @@ protected void onStart() {
     }
   }
 
+  /**
+   * This function can be called from a few different use cases, including Swiping the dialog down
+   * or calling `dismiss()` from a `BottomSheetDialogFragment`, tapping outside a dialog, etc...
+   *
+   * The default animation to dismiss this dialog is a fade-out transition.
+   * We want to utilize the swipe down dismiss animation for all cases.
+   *
+   * If this function is called from a swipe down interaction, then keep the default behavior.
+   *
+   * Else, since this is a terminal event which will finish this dialog, we override the attached
+   * {@link com.google.android.material.bottomsheet.BottomSheetBehavior.BottomSheetCallback}
+   * to call this function, after {@link BottomSheetBehavior#STATE_HIDDEN} is set.
+   * This will enforce the swipe down animation before canceling this dialog.
+   */
+  @Override
+  public void cancel() {
+    // DialogSheet was swiped down to STATE_HIDDEN, so safe to defer to parent cancel().
+    if (getBehavior().getState() == BottomSheetBehavior.STATE_HIDDEN) {
+      super.cancel();
+    } else {
+      // DialogSheet cancelled through another method.
+      WeakReference<BottomSheetDialog> dialogWeakReference = new WeakReference<>(this);
+
+      BottomSheetBehavior behavior = getBehavior();
```

Comment:
The default BototmSheetCallback that is set on the dialog will call `cancel()`. Does this still work if you just set the state to hidden, and don't override the callback at all?

Reply:
I think that would work.
However, my only concern is if the default behavior can be overridden.

I know the user can set their own `BottomSheetCallback` in the `BottomSheetBehavior`. So if they override the default callback, than we'd lose the `default implementation which calls `cancel()` on `STATE_HIDDEN` right?

Reply:
Maybe we should set an alternative way to listen? For example, `BottomSheetDialog` can have a method to add a `BottomSheetCallback` and it can delegate to that in it's own callback. That might be the clearest solution.

Or we can check if the callback that is currently set is the one from `BottomSheetDialog` and if it's not, we can wrap it then.

It's kind of broken if someone overrides the callback now though right? Because if it's not using it's callback, when the bottomsheet is hidden, the dialog won't be dismissed?

Reply:
Yeah, I think we'd need to do one of those.
I tested this out with the Catalog app in this project, and noticed that the example there is overriding the behavior itself, so the default callback is completely ignored.

<img width="917" alt="BottomSheetCallbackReferences" src="https://user-images.githubusercontent.com/1795695/60614895-7af2da00-9d93-11e9-8feb-6eda9d6a750d.png">

I'm gonna add a check to see if the set callback is the default, then to call `super.cancel()`.
If not, I'll instantiate a new one like I have currently.

Reply:
I went with option 2, of checking the currently set `BottomSheetCallback`, to see if it's the default one. If it is, I defer the logic to that callback and simply set the state to `STATE_HIDDEN`.

If it's different, then I instantiate a new callback to override the existing one to handle closing the dialog.